docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,167 @@
+
+# Motivation
+
+Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's blocks simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the if statement's blocks, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
+
+The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope. 
+
+Another benefit provided by initializers is being able to compact verbose guard clauses into a singular if statement.   
+
+Example:
+
+```lua
+function PlayerControl:GetEntities()
+    if self.RemovedSelf then
+        return self.Instances
+    end
+
+    local entity = self:TryGetEntity()
+    if entity == nil then
+        return self.Instances
+    end
+
+    local index = table.find(self.Instances, entity.Root)
+    if index == nil then
+        return self.Instances
+    end
+
+    table.remove(self.Instances, index)
+    self.RemovedSelf = true
+
+    return self.Instances
+end
+```
+
+```lua
+function PlayerControl:GetEntities()
+    if self.RemovedSelf then
+    elseif local entity = self:TryGetEntity() in entity == nil then
+    elseif local index = table.find(self.Instances, entity.Root) then
+        table.remove(self.Instances, index)
+        self.RemovedSelf = true
+    end
+
+    return self.Instances
+end
+```
+
+# Design
+
+If statements with initializers must match the below grammar. The variables declared by an initializer are only available to the if statement's blocks; any code after the if statement won't have the variables defined.
+
+```diff
+  stat = varlist '=' explist |
+      ...
+      'repeat' block 'until' exp |
+-     'if' exp 'then' block {'elseif' exp 'then' block} ['else' block] 'end' |
++     'if' cond 'then' block {'elseif' cond 'then' block} ['else' block] 'end' |
+      'for' binding '=' exp ',' exp [',' exp] 'do' block 'end' |
+      ...
++ cond = 'local' binding '=' exp ['in' exp] |
++     'local' bindinglist '=' explist 'in' exp |
++     exp
+```
+
+In the former case, the value of the first declared variable will be checked.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() then
+   print(b, "truthy block")
+else
+   print(b, "falsy block")
```

Comment:
Wanted to add that this design point is one we have to live with in perpetuity. If we decided to take this fallthrough semantics out of this RFC, then we're stuck with that behavior.

Using `boolean` is an ok example, but it doesn't strongly argue for it as opposed to some other example I posted earlier wrt `animal` and one that @vegorov-rbx posted with `pcall` example. If a design point is a "choose now or forever hold your peace," it especially needs an equally as compelling argument to support it.

So I'd rather block the RFC until we are able to agree on this design point, whether locals should fallthrough or not.

An example where the behavior is breaking change would also be good to include in the alternatives section.

```lua
if local x = f() in x % 2 == 0 then
  print(x) --> any even numbers
else
  print(x) --> nil
end
```

As is, this RFC proposes the `print(x)` in the `else` branch to print any odd numbers.